### ISSUE TITLE
DTLS support

### ISSUE THREAD

**Issue by Yamakaky** [ctz/rustls#40]
Do you plan on supporting DTLS (TLS over UDP)?

**Comment by ctz** [comment]
No plans at the minute. I know it's used for some VPNs, webrtc and some iot things. What is your use case?

**Comment by Yamakaky** [comment]
VPN ^^

**Comment by manuels** [comment]
Would you see DTLS within the scope of this library (i.e. would you accept pull request)?

**Comment by manuels** [comment]
I am currently playing around implementing DTLS. Would it be the best to do this on the tls13 branch?

**Comment by ctz** [comment]
> Would you see DTLS within the scope of this library

Perhaps. I'm not sure some of the external APIs are really suitable, to the extent that it might make sense as a separate crate built from the same codebase.

> Would it be the best to do this on the tls13 branch?

Yes. Though beware that this branch is still in major churn and is overdue a serious refactoring of `server_hs.rs`/`client_hs.rs`

**Comment by manuels** [comment]
In case you are interested: https://github.com/manuels/rustls/tree/dtls12
(although I just implemented Payload data structures and parsers up until now)

**Comment by ctz** [comment]
Looking good! 👍

**Comment by manuels** [comment]
Ok, I figured out that the situation is a bit complex, and there are several ways to implement DTLS specific stuff. Since this decision can have quite an extensive impact on the codebase I'd like to discuss how to proceed.

# The Problem
DTLS requires some additional fields in `Message` and `HandshakeMessagePayload` and `ClientHelloPayload` as well as an additional handshake message `HelloVerifyRequest`. Additionally handshakes can be fragmented on the handshake protocol level instead of the record layer level.
See the sketch where modifications are highlighted.

![dtls](https://cloud.githubusercontent.com/assets/31315/22396479/0158c544-e55b-11e6-8f60-1102e9b1f40e.png)

# Possible Solutions

## `Option`s
The additional fields could be integrated as `Option`s into the regular TLS data types and would just be ignored in the TLS implementation. E.g.

    pub struct ClientHelloPayload {
        pub client_version: ProtocolVersion,
        pub random: Random,
        pub session_id: SessionID,
        pub cookie: Option<Cookie>, // <-- !!!
        pub cipher_suites: Vec<CipherSuite>,
        pub compression_methods: Vec<Compression>,
        pub extensions: Vec<ClientExtension>,
    }

## Traits
A TLS `Message` and a DTLS `Message` are quite similar and the overlapping code could be moved into a trait `MessageTrait`. This would probably be the most rusty solution, but it would also make the code a bit more complex if most functions have a generic argument type of `MessageTrait`.

## Macros
If you want a TLS `Message` and a DTLS `Message` data type in the codebase the overlapping code could probably also be deduplicated by some macro magic. But again, it might make the codebase more complex if there are macros scattered everywhere.

Maybe you guys can up with a better idea. I would love to get some feedback to see how to continue with implementing DTLS!

**Comment by manuels** [comment]
After some time thinking about it, I think it's worth renaming the current `Message` to `TLSMessage` and creating a `Message` trait that is implemented by by `TLSMessage` and a `DTLSMessage`.

Then all struct `impl`s just need to be `impl<Msg: Message+Codec> MyStruct`. Any comments on this?

**Comment by ctz** [comment]
DTLS/TLSMessage sounds ok, but beware that some performance relies on Message being moved by value into the processing functions: so the traits will need to cover these cases by exposing (eg) `take_payload`.

sorry, i overlooked your earlier comment. 

for `ClientHelloPayload` and the like, are the functions which process these likely to be mostly shared between TLS and DTLS? if they're vastly different, then having a ClientHelloPayloadDTLS subtype would suffice (like some of the TLS1.3 types). otherwise, optional elements would be ok.

**Comment by manuels** [comment]
> for ClientHelloPayload and the like, are the functions which process these likely to be mostly shared between TLS and DTLS?

They are basically the same. The DTLS version of ClientHelloPayload just adds an optional cookie.

**Comment by manuels** [comment]
Hey @ctz, I've seen in your TODO project that you are refactoring the state machine.
Do you already have some WIP that you could push to a new branch to github, so I can see where this is going.
I am at the point that I'd have to adapt it to implement DTLS.
Would be great!

Thanks!

**Comment by ctz** [comment]
I don't have any wip branch for that yet. But I'll push one once I do and link it here.

**Comment by manuels** [comment]
In case this is useful: I tried to summarize the TLS state machine a while a go ![TLS state machine](https://github.com/manuels/tls-state-machine/blob/master/state_machine.svg).

**Comment by manuels** [comment]
A quick update. I had to refactor the TLS specific code of `SessionCommon` into a `StreamTransport` struct. I will now proceed and implement a `DatagramTransport` struct with a similar but DTLS specific behaviour.

**Comment by manuels** [comment]
I am wondering how the user should decide which protocol to use.

Either the instantiation of `{Client,Server}Session` takes a Trait parameter ala `ClientSession<TLS>` (or `ClientSession<DTLS>`) at compile time, or it will decided at runtime using the ProtocolVersion.
But right now I do not see, how to implement this without a lot of `ClientSessionImpl` -> `ClientSessionImpl<M:Message>` renaming in the code.

@ctz, do you have a comment on this?

**Comment by ctz** [comment]
How about separate `DatagramClientSession`/`DatagramServerSession` types? I'm assuming these types need some kind of extra methods for timed triggering of retransmits?

On the question of `ClientSessionImpl`, could we inject implementations of a trait that deals with the important differences between datagram/stream TLS? Like which type of message to use. Actually those traits could be quite big. Hm.

**Comment by opensourcegeek** [comment]
Just wondering if there has been any progress on this issue, I'm working on IoT and support for DTLS would be very handy.

**Comment by manuels** [comment]
Wireguard solved my problems so far and I do not have time to work on DTLS currently.
Maybe that might change in a year's time or so...

**Comment by henriktorget** [comment]
@manuels Do you have time to have a look at it now? I would love to see some progress on this implementation.

**Comment by TotalKrill** [comment]
A DTLS implementation would be very awesome, I hope you find some time @manuels, seems like you were very close!

**Comment by TimonPost** [comment]
It would be great to see DTLS support, is there any progress?

**Comment by ZizhengTai** [comment]
Would really love to see DTLS implemented in Rustls. We need it for some WebRTC work and currently are forced to use OpenSSL.

**Comment by ctz** [comment]
Some plans for this, but have been waiting for DTLS1.3. That is nearing completion now.

**Comment by repi** [comment]
Any additional thoughts or updates on this? Something we've been interested in as well and investigating options for, cc @luna-duclos

**Comment by lachlansneff** [comment]
A pure-rust DTLS implementation would be fantastic for embedded usecases. Any further updates on this?

**Comment by djc** [comment]
DTLS 1.3 looks like it is nearing completion, currently in the RFC editor's queue.

Meanwhile, for some use cases maybe QUIC + its datagram extension is a decent fit?

(If someone is interested in sponsoring DTLS 1.3 support, I'd be happy to connect.)

**Comment by maufl** [comment]
I recently discovered https://github.com/webrtc-rs/dtls but I don't know whether it is already usable.

**Comment by summit-solutions** [comment]
Also since QUIC has progressed very well, which runs over UDP but uses TLS building blocks from the regular TCP TLS implementations, the requirement for DTLS (UDP) has ... well, diminished.

**Comment by lachlansneff** [comment]
> I recently discovered https://github.com/webrtc-rs/dtls but I don't know whether it is already usable.

Unfortunately, that crate seems to be very tightly fused to std.

**Comment by markmandel** [comment]
> I recently discovered https://github.com/webrtc-rs/dtls but I don't know whether it is already usable.

There are is a [e2e test PR](https://github.com/webrtc-rs/dtls/pull/2) that shows it in action, but there seem to be a few structs that are missing. Not sure what is happening there.

**Comment by djc** [comment]
Folks, it's useful to have a link to that crate here but further discussion about it seems off-topic for this issue.

**Comment by ctz** [comment]
My feeling is that DTLS would need a totally different public API to rustls, to the extent that it doesn't really make sense to have it in the same crate (even as a cargo feature). What I would be open to is refactoring the core crate to allow reuse of the parts common between TLS/DTLS,. What I would never want to see in the core crate is stuff like https://github.com/openssl/openssl/blob/54b40531307fcaba1206e98f4cae73f0532fbdbb/ssl/record/ssl3_buffer.c#L45-L48 -- that would seem to me to be evidence of bad architecture.

As for actually making progress on this, the tlswg continue to alter the design, and I'm not sure writing a DTLS1.2-only implementation is a sensible idea at this point.

The other thing I'm aware of is that DTLS and embedded use cases often come hand-in-hand, so the DTLS crate would likely have to pay more attention to the constraints this brings.

**Comment by ShadowJonathan** [comment]
While playing around with DTLS, I came to the conclusion that TLS and DTLS are working with two entirely different kinds of streams, and that knowledge about MTU, framing, and other items are important to a consuming DTLS (de)encryptor.

Thus, I concur this has to be a different API/crate, if it means it could consume a std `UdpSocket`, and correctly abstract it's framed `send` and `recv` over it.

**Comment by ShadowJonathan** [comment]
For anyone curious about this development, i'm going to attempt to create an own implementation of this, as per https://github.com/ruma/lb/issues/7, taking https://github.com/pion/dtls and this library as inspiration.

I know i'll be making mistakes, so i'm inviting people who are interested in this to help me out as well.

If there are any rustls maintainers reading this; Collaboration is appreciated, once/if i get the implementation up and close to something usable, i'm offering the repo to be adopted by the rustls org, including crates.io ownership and all.

I'll need recommendations on supported cipher suites, though for now it is a goal to be compatible with at least [some of these](https://github.com/pion/dtls/blob/415eb6f8a18b35b8fbae137c83046692e159f1e6/internal/ciphersuite/ciphersuite.go#L48-L67) cipher suites.

I'll initially be executing a rough API design as described [here](https://github.com/ruma/lb/issues/7#issuecomment-986027017), though obviously it's open for discussion.

Efforts are going to be [here](https://github.com/ShadowJonathan/dtls-rs), though with me doing this in my free time, commits are going to be sporadic, help is appreciated.

**Comment by Ralith** [comment]
Just off-hand, I'd highly recommend *not* trying to wrap a `UdpSocket`, but instead exposing methods to consume/write out datagrams. That will make testing simpler and less fragile, and make the interface more broadly applicable.

**Comment by djc** [comment]
Agreed with @Ralith, would be nice to mimic the rustls API where possible. I might try reviewing your work as you go, if you want to tag me in PRs or something.

**Comment by ShadowJonathan** [comment]
One of my goals is definitely to be as modular as possible to the underlying transport(s), i.e. from the rough API draft, "`DTLSSessionMachine`" already looks quite similar to `ConnectionCommon`, though i'll have to think through and study the `rustls` api surface before i can commit to what i'm thinking right now.

**Comment by ShadowJonathan** [comment]
@djc @ctz I'd like to collaborate a bit more closely on this (especially with the idea of such a core crate), so i can quickly ask some questions in the future, do y'all have any preferred IM handle i can contact you with? I'd personally prefer matrix: [`@jboi:jboi.nl`](https://matrix.to/#/@jboi:jboi.nl), my twitter DMs are also open.

**Comment by djc** [comment]
Feel free to hit me up on Matrix or Twitter. On Matrix I'm @djc:mozilla.org I think.

**Comment by ShadowJonathan** [comment]
(I know its been 3 years, but I'd like to mention that my motivation has dropped for this feature)

**Comment by bzted** [comment]
Has there been more discussion about this topic? I was thinking of implementing DTLS 1.3 on rustls as its current trait based architecture might make it feasible.

**Comment by djc** [comment]
> Has there been more discussion about this topic?

Nothing substantial AFAIK!

**Comment by algesten** [comment]
Worth mentioning, since I started following this issue a while back, I decided earlier this year to write my own DTLS 1.2 impl ([dimpl](https://crates.io/crates/dimpl)).  The problem I face is that WebRTC is stuck on DTLS 1.2 (needed for [str0m](https://crates.io/crates/str0m)), and I don't know how much appetite there is in rustls for that narrow subproblem specifically.

At the time, I didn't know DTLS well enough to feel confident I could contribute to rustls. That has changed a bit now. Ideally I'd drop the dimpl crate in favor of rustls, and if there is some interest in doing DTLS 1.2 (and 1.3) in rustls, I'd be happy to collaborate on that.

**Comment by djc** [comment]
[Previously](https://github.com/rustls/rustls/issues/40#issuecomment-961391105):

> My feeling is that DTLS would need a totally different public API to rustls, to the extent that it doesn't really make sense to have it in the same crate (even as a cargo feature). What I would be open to is refactoring the core crate to allow reuse of the parts common between TLS/DTLS,. What I would never want to see in the core crate is stuff like https://github.com/openssl/openssl/blob/54b40531307fcaba1206e98f4cae73f0532fbdbb/ssl/record/ssl3_buffer.c#L45-L48 -- that would seem to me to be evidence of bad architecture.
> 
> As for actually making progress on this, the tlswg continue to alter the design, and I'm not sure writing a DTLS1.2-only implementation is a sensible idea at this point.
> 
> The other thing I'm aware of is that DTLS and embedded use cases often come hand-in-hand, so the DTLS crate would likely have to pay more attention to the constraints this brings.

I think the crypto provider API might make for a good base layer if you wanted to do a DTLS 1.2 implementation?

**Comment by bzted** [comment]
> Worth mentioning, since I started following this issue a while back, I decided earlier this year to write my own DTLS 1.2 impl ([dimpl](https://crates.io/crates/dimpl)). The problem I face is that WebRTC is stuck on DTLS 1.2 (needed for [str0m](https://crates.io/crates/str0m)), and I don't know how much appetite there is in rustls for that narrow subproblem specifically.
> 
> At the time, I didn't know DTLS well enough to feel confident I could contribute to rustls. That has changed a bit now. Ideally I'd drop the dimpl crate in favor of rustls, and if there is some interest in doing DTLS 1.2 (and 1.3) in rustls, I'd be happy to collaborate on that.

Right now I'm only interested in DTLS 1.3, as I'm working on experimental PQC variants of TLS on rustls (https://github.com/bzted/rustls). Would you be interested in implementing only DTLS 1.3?

**Comment by tgeoghegan** [comment]
I'm dusting off this old issue to restart the discussion on whether and then how to add support for Datagram TLS (DTLS) in rustls. DTLS is used in performance sensitive situations where TCP or QUIC's transport-layer delivery guarantees are not needed or wanted, instead letting the application manage retries and ordering. Examples include multiplayer games or [WebRTC](https://en.wikipedia.org/wiki/WebRTC).

Adding DTLS could unblock adoption of rustls by applications that need both TLS and DTLS, and are reluctant to depend on two different implementations.

My employer ([Internet Security Research Group](https://abetterinternet.org)) is prepared to contribute engineering time to this project. To that end, I've been working for some weeks now on adding DTLS to rustls. This has helped inform my proposed design for it, and I also believe I have some pieces of the implementation that could start landing soon if we have consensus. I expect to spend quite a bit of time in the coming months continuing implementation and working with rustls maintainers to land the work.

I'll be commenting in this issue with some detailed discussion of how I think rustls could add DTLS support, what kind of API rustls might expose, and sketches of a validation plan. I welcome any and all discussion on whether this is a good idea, how to do it, or how it should land in rustls. I'll do my best to answer questions and discuss, but of course the decisions will be made by the rustls maintainers. And please let me know if it would be preferable to start a new issue, since much of the backscroll here is in the context of older versions of the library.

**Comment by tgeoghegan** [comment]
DTLS 1.2 and 1.3 are designed to reuse message definitions and state machines from the corresponding TLS versions. That means they can be implemented by adding optional fields to existing structures or conditional checks to existing TLS 1.2 and 1.3 state machine transitions without excessive disruption.

This document lays out a plan for doing this implementation and highlights some areas that need more thought and input from rustls maintainers. Much of this plan is tentative: we won't know for sure that it works until we actually try, and the plan gets shakier as we get to the parts I haven't spent time implementing yet.

DTLS support will be brought exclusively to the current `main` branch of rustls, targeting version 0.24 and onward.

## Implementation status

I have a _very_ rough work in progress branch [on GitHub](https://github.com/tgeoghegan/rustls/tree/dtls-messages) where I am exploring the rustls codebase and validating that my implementation strategy will work. Obviously it needs a lot of work before it can be reviewed, and I'd chunk it up into digestable pull requests.

## DTLS messages and features

The point of DTLS is provide strong security guarantees to datagram oriented transports, useful for applications where performance requirements rule out the overhead of TCP and which can tolerate out of order or lost messages. But awkwardly, the TLS handshake still has ordering and reliability requirements. To address this, Datagram TLS' record layer adds the following features:

- epoch and sequence numbers to allow out of order transmission and replay detection;
- handshake message fragmentation and reassembly;
- a state machine for retransmission of messages until acked by recipients.

Additionally, DTLS 1.2 and 1.3 both rely on [RFC 9146's `connection_id` extension](https://datatracker.ietf.org/doc/html/rfc9146) for servers to identify the appropriate security context for inbound traffic.

### Handshake message fragmentation

Handshake message fragmentation and reassembly has implications for the send and receive paths (`conn::send::SendPath` and `conn::receive::ReceivePath`, respectively).

Message fragmentation will be implemented in `conn::send::SendPath` and `msgs::fragmenter::MessageFragmenter`, transforming `Message` structures into byte vectors stored in `SendPath::sendable_tls` for eventual transmission. Those objects are already aware of which protocol version they're running, so adding conditional logic for DTLS 1.2 or 1.3 is straightforward.

Currently, `MessageFragmenter` will transform an `EncodedMessage<Payload>` into an iterator yielding `EncodedMessage<OutboundPlain>`, where `OutboundPlain` contains an opaque byte string. This interface doesn't work for DTLS, because rather than an opaque byte string, each fragment needs to be a `struct DTLSHandshakeFragment`, retaining the original message's `msg_type` and `length` and containing appropriate `message_seq, fragment_offset` and `fragment_length` fields ([RFC 9147 section 5.2](https://datatracker.ietf.org/doc/html/rfc9147#section-5.2)). To facilitate this, we'll add an interface to `MessageFragmenter`:

```rust
/// Take a DTLS handshake message and fragment it into multiple unencrypted outbound messages,
/// each consisting of a DTLSPlaintext ([1]). Other DTLS messages may not be fragmented.
///
/// [1]: https://datatracker.ietf.org/doc/html/rfc9147#appendix-A.1
fn fragment_dtls_handshake_message<'a>(
    &self,
    epoch_and_sequence: EpochAndSequence,
    msg_type: HandshakeType,
    handshake_sequence_number: u16,
    handshake_payload: &'a [u8],
) -> impl Iterator<Item = EncodedMessage<DtlsHandshakeFragment<'a>>>
```

Another wrinkle is handshake message flights: when the transport is connection oriented like TCP or QUIC, it's straightforward to concatenate handshake messages into flights and then send the resulting byte stream to the TCP layer which will pack those messages into one or more packets. But with DTLS on UDP, we must pack multiple handshake messages into a single datagram at the DTLS layer. Thus, we also add an interface:

```rust
/// Fragment the provided flight of handshake messages (represented as tuples of the handshake
/// type and the encoded handshake message payload) as one or more DTLS records.
///
/// Each record may contain more than one handshake message, or just a single handshake
/// fragment.
///
/// `handshake_sequence_number` is the handshake sequence number for the first message in this
/// flight.
 fn fragment_dtls_handshake_message_flight<'a>(
    &self,
    epoch_and_sequence: EpochAndSequence,
    handshake_sequence_number: u16,
    handshake_messages: &'a [(HandshakeType, Vec<u8>)],
) -> Vec<EncodedMessage<Payload<'a>>>
```

`rustls::conn::send::SendPath::send_msg` will use these new interfaces to appropriately fragment and transmit DTLS messages.

#### Implementation status

Message fragmentation is in good shape! I've got it working for DTLS 1.2 and 1.3 and have added a good amount of test coverage for various edge cases.

### Message reassembly (deframing)

Correspondingly, message reassembly will be implemented in `conn::receive::ReceivePath` and `msgs::deframer::Deframer`. Besides the obvious differences in message encoding and how multiple handshake messages can be packed into a single DTLS record, there's an additional problem that the DTLS receive side has to contend with: messages can arrive out of order. This means the existing deframing machinery that consumes ordered bytes from TCP can't be used as-is. We add interfaces to `msgs::deframer::Deframer`:

```rust
/// Input a DTLS record containing one or more handshake fragments so that they can be
/// re-ordered and re-assembled by [`Self::coalesce_dtls`]. There should not be any trailing
/// bytes on the message payload.
///
/// `msg` is a parsed TLS record, which may contain one or more handshake messages, each
/// starting with a handshake header.
///
/// `bounds` is the position within the containing buffer of the record payload. That is, it
/// begins at the start of the first handshake header.
pub(crate) fn input_message_dtls(
    &mut self,
    msg: EncodedMessage<&'_ [u8]>,
    bounds: Range<usize>,
) -> Result<(), Error>;

 /// Coalesce the contents of `containing_buffer` into one or more complete DTLS handshake
/// messages.
///
/// `containing_buffer` is understood to contain some number of DTLS records containing
/// handshake messages, i.e., a record header, then one or more handshake headers and payloads.
/// Before calling this function, each of those records must have been parsed by
/// [`Self::deframe`] and then input into this deframer with [`Self::input_message_dtls`].
///
/// If `containing_buffer` contains all the fragments of a handshake message, then on return,
/// the buffer will contain the coalesced (reassembled) handshake message, followed by any
/// remaining uncoalesced fragments.
///
/// If `containing_buffer` contains all the fragments of multiple handshake messages, then on
/// return, the buffer will contain coalesced handshake messages, ordered by the handshake
/// sequence number, not to be confused with the sequence number at the DTLS record layer.
///
/// Coalesced handshake messages consist of the handshake header of the first fragment,
/// concatenated with just the handshake payloads of subsequent fragments. Coalesced messages
/// include `DTLSHandshake.{message_seq, fragment_offset, fragment_length}` values but these are
/// no longer meaningful since the message is coalesced. See [1], [2] for details of the
/// `DTLSHandshake` structure.
///
/// After calling this method, callers should call [`Self::complete_span`] to find out the
/// position of the next coalesced handshake message, if any, and then [`Self::message`] to
/// obtain it.
///
/// More fragments may then be added into the deframer by calling [`Self::deframe`] and
/// [`Self::input_message_dtls`] again.
///
/// [1]: https://datatracker.ietf.org/doc/html/rfc6347#section-4.2.2
/// [2]: https://datatracker.ietf.org/doc/html/rfc9147#section-5.2
pub(crate) fn coalesce_dtls(&mut self, containing_buffer: &mut [u8]);
```

DTLS messages on the wire are not exactly the same as TLS ones (they contain epoch and sequence numbers or fragment offsets), but the handshake state machine can mostly be insulated from this, as these differences mostly appear in `crypto::cipher::messages::EncodedMessage` structures and need not be reflected in `msgs::Message` (which represents a message *after* it's been reassembled from fragments).

#### Implementation status

Message deframing is also in good shape with good test cases for DTLS 1.2 and 1.3 messages, including various out of order message cases. There are some TODOs like using a fixed size buffer for copying message fragments around during coalescing so that we don't make strong assumptions about the PMTU in use.

### Replay detection and epoch handling

DTLS records include epoch and sequence numbers. Sequence increments with each message sent while epoch changes when key material changes ([RFC 9147 Section 6.1](https://datatracker.ietf.org/doc/html/rfc9147#section-6.1)). Since transports like UDP guarantee neither delivery, only-once delivery or in-order delivery, the receive path must:

- discard messages from old epochs;
- buffer messages from future epochs;
- keep track of a window of observed sequence numbers in the current epoch to discard replays.

This requires keeping track of the current epoch and sequence (which is necessary for deprotecting records anyway, since those values are incorporated into cipher AADs) and somewhat subtle interactions between `rustls::msgs::Deframer` and `conn::receive::ReceivePath` so that replays or tardy messages can be gracefully discarded.

#### Implementation status

Not yet done. Much of the plumbing is there, but lots of test cases are needed to get confidence here.

### Message encoding and decoding

For backward compatibility reasons, TLS 1.3 messages will sometimes claim to be TLS 1.0 or 1.2, depending on context ([RFC 8446, section 5.1](https://datatracker.ietf.org/doc/html/rfc8446#section-5.1)). This complicates implementations because outgoing messages are not self-describing. DTLS 1.3, unfortunately, makes this worse, because it uses the same backward compatibility tricks ([RFC 9147, section 4](https://datatracker.ietf.org/doc/html/rfc9147#section-4)) and also introduces the unified header, which is used instead of a full record header unless the message is unencrypted.

The upshot is that contexts like `crypto::cipher::messages::EncodedMessage<OutboundOpaque>::encode` or `crypto::cipher::MessageEncrypter::encrypt` need to know what protocol is in use and occasionally extra bits of contextual information (such as whether the message is an initial or retried client hello) to make appropriate encoding decisions.

To address this, we introduce `struct EncodingContext`, a structure that captures that contextual information. It is stored in a field on `OutboundOpaque` so that `EncodedMessage<OutboundOpaque>` values can be passed around various rustls components while preserving the information needed to eventually encode messages onto the wire.

#### Implementation status

Much of this is done. DTLS 1.2 records are encoded and decoded, and DTLS 1.3 records using full headers (for unencrypted ClientHellos) or unified headers are supported as well. More testing is needed, though, especially against other implementations: it doesn't mean all that much that rustls can talk to itself.

### Message retransmission state machine

The retransmission state machine gets into some public API design discussion. rustls exposes connection-oriented API for TLS (`client::connection::ClientConnection` and `client::connection::ServerConnection`), both of which consist of `conn::ConnectionCommon` and `conn::ConnectionCore` and finally `common_state::CommonState`, which contains `conn::send::SendPath` and `conn::send::ReceivePath`. None of these objects actually send or receive bytes over something like a socket but rather will accumulate incoming bytes into `ConnectionCommon::buffers` and outgoing ciphertext in `SendPath::sendable_tls`. Then some other component that's actually responsible for interacting with the network is meant to periodically check for outgoing ciphertext and do something with it like write it to a `TcpSocket`. An example of this is `rust-util::stream::Stream`, which can do so with an arbitrary `std::io::Write` implementer.

I think that an analogous construction for UDP is the appropriate place to implement the retransmission state machine. The tricky thing is that we don't want to re-use this entire stack of connection abstractions or `Stream`. Everything from `ConnectionCommon` upward in the stack is designed around `std::io::{Write,Read}`. This is reasonable for TLS or QUIC connections which are oriented around streams, but is an awkward fit for datagram protocols. Concretely, Rust's `std::net::UdpSocket` deliberately does not implement the `std::io` traits, instead exposing a libc-inspired pair of `send, recv` methods.

My objective is to avoid exposing a DTLS API based on `std::io::{Read,Write}` so that users don't have to awkwardly implement those traits on whatever actual network transport they're using.

If we inspire ourselves from the layout of the QUIC implementation, we would add `rustls::datagram::{ClientSocket, ServerSocket}`. These `Socket` types are where the DTLS retransmission state machine will be implemented. They will be generic over a trait:

```rust
trait UdpSocketLike {
    type Error: std::error::Error;

    fn send<B: AsRef<[u8]>>(&mut self, buf: B) -> Result<usize, Self::Error>;

    fn recv<B: AsMut<[u8]>>(&mut self, buf: B) -> Result<usize, Self::Error>;
}
```

...so that they can be used with an `std::net::UdpSocket` or something else. We'll provide `impl UdpSocketLike for std::net::UdpSocket`. We could also provide a blanket implementation `UdpSocketLike for T where T: std::io::Write + std::io::Read`.

This layer is where I want to implement transmission timeouts and retransmission ([RFC 9147 section 5.8](https://datatracker.ietf.org/doc/html/rfc9147#section-5.8)). This is the component that's responsible for actually pushing and pulling bytes on and off the network, so it feels like a better place to do retries and deal with the sordid details of I/O than either the state machines in `client::hs` or `server::hs` or the send and receive paths (`conn::send::SendPath`, `conn::receive::ReceivePath`).

#### Alternatives to consider

- Instead of putting retransmission out in new `ClientSocket, ServerSocket` types, we could change the contract of `ChunkVecBuffer` so that instead of taking ownership of a chunk by calling `ChunkVecBuffer::take_one_vec` or `::pop`, you'd have to explicitly indicate that you successfully transmitted a chunk by calling another method. That would let `ChunkVecBuffer` implement the retransmission state machine: `take_one_vec` and `pop` would hand out the same chunk until the consumer layer indicates it had successfully send the chunk.
- We could decide that it's perfectly reasonable to implement `std::io::{Read, Write}` on `std::net::UdpSocket` and then re-use much more of the existing rustls connection machinery.

#### Implementation status

I have a lot of this implemented, which has been useful so far to do high level tests where I drive the entire handshake process and exchange `ApplicationData` between client and server. However I think a lot more work needs to be done on API design. I also think that some or many of the items in [#2400](https://github.com/rustls/rustls/issues/2400) will have an impact here, because some of the `Connection` primitives or other layers of crate internals may be refactored significantly on the way to rustls 0.24.

### DTLS state machine

There's also some changes for DTLS relative to the TLS state machine:

- DTLS 1.3 endpoints MUST NOT send `ChangeCipherSpec` messages
- handshakes may begin with a stateless cookie exchange intended to mitiate denial of service attacks
- DTLS 1.3 endpoints must send ACK messages in response to handshake messages

These requirements mean that the TLS state machines (`client::hs`, `server::hs`) will have to become aware of DTLS and DTLS version, much as they already conditionalize their behavior based on whether TLS 1.2, 1.3 or QUIC is in use.

We'll add a variant `rustls::common_state::Protocol::Udp` to indicate that UDP and hence DTLS are in use. The 1.3 state machine already has a `Protocol` value plumbed into its states (actually, it's a field of a `HandshakeState` value). The 1.2 state machines will have `HandshakeState` threaded through them so that state transitions can similarly adjust for DTLS as needed.

#### Implementation status

A lot of the plumbing necessary for the state machines to make decisions based on protocol version is in place, but much of this behavior is not yet implemented. I think the most important test here will be to try it out against another DTLS implementation (especially within Bogo).

## Cryptography

DTLS cryptography is largely the same as regular TLS. The major difference is that the 64 bit sequence number used as the AEAD nonce (TLS 1.3) or input to the MAC (TLS 1.2) is now actually a 16 bit epoch and a 48 bit sequence number. The cryptography layer already keeps track of sequence number in `rusts::crypto::cipher::record_layer::{EncryptionState,DecryptionState}`, so they can be taught to deal with DTLS epoch and sequence.

For DTLS 1.3, we also need to implement [record sequence number encryption](https://datatracker.ietf.org/doc/html/rfc9147#section-4.2.3).

#### Implementation status

The plumbing to make the crypto layer aware of what protocol is in use is there, but the special logic for constructing AADs from DTLS epoch and sequence needs work, as does record sequence number encryption.

## DTLS API considerations

### Refusal to fragment messages besides handshakes and PMTU management

The DTLS specifications allow for an implementation to split messages besides handshakes into multiple records. For instance, [RFC 9147 Section 10](https://datatracker.ietf.org/doc/html/rfc9147#section-10) reads:

> Application data messages are carried by the record layer and are split into records and encrypted based on the current connection state. The messages are treated as transparent data to the record layer.

However, it seems unhelpful for the DTLS layer to split application layer messages, since it's the application layer that will have to make decisions about ordering, reassembly and what to do about dropped messages. So the implementation will refuse to fragment application data messages that are too large for the configured maximum fragment size.

The specifications make it clear that that the DTLS and application layers will need to agree on the Path Minimum Transmission Unit (PMTU) in use ([RFC 9147 section 4.4](https://datatracker.ietf.org/doc/html/rfc9147#section-4.4)). rustls already allows clients to configure the maximum fragment size (`ClientConfig::max_fragment_size`), so it seems more reasonable for rustls to require clients to send application data in chunks smaller than that configured size. That way, if the application layer protocol does have some kind of sequence numbering of its own, it can be encoded into the plaintext before DTLS records are constructed or encrypted.

## Dependencies

Implementing DTLS should not add new dependencies to rustls.

## Testing plan

We'll write straightforward Rust tests for any new items that get added (see for instance the `deframer::dtls_test` and `fragmenter::dtls_test` modules I have in my working branch), and API level tests for new datagram oriented public API (per [rustls testing guidelines](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md#testing)).

But this doesn't suffice for validating a TLS implementation. The first step will be to enable the DTLS tests in [bogo](https://github.com/rustls/rustls/tree/main/bogo).

- Does bogo adequately exercise DTLS 1.2 and 1.3?
- What other validation or testing can we do?
- Are there other test suites from other (D)TLS implementations we can import into rustls?

## Future work/open questions

- Encrypted Client Hello (RFC 9849) is supported for DTLS 1.3 only. Given the modest adoption of ECH online (on top of DTLS 1.3 itself not being very popular), is it worth implementing?
- Should we have all this gated on a new cargo feature?
    - Relatedly, should we somehow mark this work as experimental until it's been vetted sufficiently?
- What version of rustls should this target?

## References

[RFC 6347 (DTLS 1.2)](https://datatracker.ietf.org/doc/html/rfc6347)
[RFC 9147 (DTLS 1.3)](https://datatracker.ietf.org/doc/html/rfc9147)
[RFC 9849 (Encrypted Client Hello)](https://datatracker.ietf.org/doc/html/rfc9849)

**Comment by algesten** [comment]
> Adding DTLS could unblock adoption of rustls by applications that need both TLS and DTLS, and are reluctant to depend on two different implementations.

A quick note. Since I wrote my above comment, I'd had help to implement DTLS 1.3 in addition to the 1.2 I original made. That obviously doesn't change your point about wanting the same TLS library for both TCP and UDP, but I figured it's worth [mentioning dimpl](https://github.com/algesten/dimpl) has gotten a bit further along.

**Comment by djc** [comment]
Here are my thoughts on your notes.

> DTLS 1.2 and 1.3 are designed to reuse message definitions and state machines from the corresponding TLS versions. That means they can be implemented by adding optional fields to existing structures or conditional checks to existing TLS 1.2 and 1.3 state machine transitions without excessive disruption.

Note that instead of having runtime variance, it would be preferable where possible to instead implement this in a type-safe way, like potentially injecting types (particular private ones) with a generic argument to convey DTLS vs TCP/QUIC. Obviously this won't be practical everywhere but I wanted to mention it upfront -- where we can avoid run-time optionality/conditional checks, it would be good to do so.

This also helps with our code size goals, where we strive to allow dead code elimination during compilation/linking to drop unused code -- so most DTLS code should not appear in binaries that don't use it.

> I have a _very_ rough work in progress branch [on GitHub](https://github.com/tgeoghegan/rustls/tree/dtls-messages) where I am exploring the rustls codebase and validating that my implementation strategy will work. Obviously it needs a lot of work before it can be reviewed, and I'd chunk it up into digestable pull requests.

Not sure how @ctz feels about it, but I'd recommend building it up in smaller PRs on `main` rather than building up a parallel branch to land in the longer-term future. This is both to facilitate early feedback on the lower-level parts (which could potentially then spiral out into large changes needed in higher-level parts) and on the process of chunking up into commits that are digestible.

If we're worried that we might want to rip stuff out before release let's figure out alternative ways to do so. I don't think waiting for a long term before merging anything is going to be helpful.

> Message fragmentation will be implemented in `conn::send::SendPath` and `msgs::fragmenter::MessageFragmenter`, transforming `Message` structures into byte vectors stored in `SendPath::sendable_tls` for eventual transmission. Those objects are already aware of which protocol version they're running, so adding conditional logic for DTLS 1.2 or 1.3 is straightforward.

Note that I think `sendable_tls` is going to go away in favor of external buffers in 0.24. For DTLS, I'd aim for using external buffers from the start.


> Correspondingly, message reassembly will be implemented in `conn::receive::ReceivePath` and `msgs::deframer::Deframer`. Besides the obvious differences in message encoding and how multiple handshake messages can be packed into a single DTLS record, there's an additional problem that the DTLS receive side has to contend with: messages can arrive out of order. This means the existing deframing machinery that consumes ordered bytes from TCP can't be used as-is. We add interfaces to `msgs::deframer::Deframer`:

Not sure it makes sense to use the existing deframer for this; at least, we'll want to make sure we don't add (much) complexity to the TCP/QUIC path that is only used for DTLS.

> This requires keeping track of the current epoch and sequence (which is necessary for deprotecting records anyway, since those values are incorporated into cipher AADs) and somewhat subtle interactions between `rustls::msgs::Deframer` and `conn::receive::ReceivePath` so that replays or tardy messages can be gracefully discarded.

Again, this sounds like mayb eresuing the existing `Deframer` doesn't make sense (and maybe not even `ReceivePath`?).

> If we inspire ourselves from the layout of the QUIC implementation, we would add `rustls::datagram::{ClientSocket, ServerSocket}`. These `Socket` types are where the DTLS retransmission state machine will be implemented. They will be generic over a trait:
> 
> trait UdpSocketLike {
>     type Error: std::error::Error;
> 
>     fn send<B: AsRef<[u8]>>(&mut self, buf: B) -> Result<usize, Self::Error>;
> 
>     fn recv<B: AsMut<[u8]>>(&mut self, buf: B) -> Result<usize, Self::Error>;
> }
> 
> ...so that they can be used with an `std::net::UdpSocket` or something else. We'll provide `impl UdpSocketLike for std::net::UdpSocket`. We could also provide a blanket implementation `UdpSocketLike for T where T: std::io::Write + std::io::Read`.

I think we should avoid this in favor of a sans-I/O approach, where a DTLS client or server just takes `&[u8]` (or `TlsInputBuffer`-like) input and writes output into `&mut Vec<u8>` (or a trait implementation exposing similar primitives). This is the direction TCP-based rustls is going too for 0.24. (Note that `Stream` has moved to rustls-util.)

> This layer is where I want to implement transmission timeouts and retransmission ([RFC 9147 section 5.8](https://datatracker.ietf.org/doc/html/rfc9147#section-5.8)). This is the component that's responsible for actually pushing and pulling bytes on and off the network, so it feels like a better place to do retries and deal with the sordid details of I/O than either the state machines in `client::hs` or `server::hs` or the send and receive paths (`conn::send::SendPath`, `conn::receive::ReceivePath`).

Timeouts and retransmission should be wrapped around the inner TLS state.

> I have a lot of this implemented, which has been useful so far to do high level tests where I drive the entire handshake process and exchange `ApplicationData` between client and server. However I think a lot more work needs to be done on API design. I also think that some or many of the items in [#2400](https://github.com/rustls/rustls/issues/2400) will have an impact here, because some of the `Connection` primitives or other layers of crate internals may be refactored significantly on the way to rustls 0.24.

What is the base of your branch?

> ## Testing plan
> 
> We'll write straightforward Rust tests for any new items that get added (see for instance the `deframer::dtls_test` and `fragmenter::dtls_test` modules I have in my working branch), and API level tests for new datagram oriented public API (per [rustls testing guidelines](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md#testing)).
> 
> But this doesn't suffice for validating a TLS implementation. The first step will be to enable the DTLS tests in [bogo](https://github.com/rustls/rustls/tree/main/bogo).
> 
> * Does bogo adequately exercise DTLS 1.2 and 1.3?
> * What other validation or testing can we do?
> * Are there other test suites from other (D)TLS implementations we can import into rustls?

I think answering these questions should be prioritized.

> ## Future work/open questions
> 
> * Encrypted Client Hello (RFC 9849) is supported for DTLS 1.3 only. Given the modest adoption of ECH online (on top of DTLS 1.3 itself not being very popular), is it worth implementing?

I don't think it should be part of the initial implementation effort.

> * Should we have all this gated on a new cargo feature?

If it doesn't need additional dependencies, probably not.

> * Relatedly, should we somehow mark this work as experimental until it's been vetted sufficiently?

Probably makes sense to put some warnings in high-level API docstrings.

> * What version of rustls should this target?

I think you answered this above? Probably 0.24 (but not necessarily 0.24.0).

**Comment by tgeoghegan** [comment]
Thank you for the swift and thoughtful reply, Dirkjan.

> Note that instead of having runtime variance, it would be preferable where possible to instead implement this in a type-safe way, like potentially injecting types (particular private ones) with a generic argument to convey DTLS vs TCP/QUIC. Obviously this won't be practical everywhere but I wanted to mention it upfront -- where we can avoid run-time optionality/conditional checks, it would be good to do so.

Yes, this is a good design goal, and now that I think about it further, conditionalizing behavior on generic types is going to be cleaner in many or most places than wiring in miscellaneous booleans and `if protocol.is_dtls() {} else {}` checks.

> Not sure how @ctz feels about it, but I'd recommend building it up in smaller PRs on `main` rather than building up a parallel branch to land in the longer-term future. This is both to facilitate early feedback on the lower-level parts (which could potentially then spiral out into large changes needed in higher-level parts) and on the process of chunking up into commits that are digestible.
> 
> If we're worried that we might want to rip stuff out before release let's figure out alternative ways to do so. I don't think waiting for a long term before merging anything is going to be helpful.

Sure. What I'm hearing is that we want to err on the side of landing code and maintaining momentum, which I definitely support. 

> Note that I think `sendable_tls` is going to go away in favor of external buffers in 0.24. For DTLS, I'd aim for using external buffers from the start.

Ack. I will study the recent work on external buffers and adapt the implementation accordingly.
 
> Not sure it makes sense to use the existing deframer for this; at least, we'll want to make sure we don't add (much) complexity to the TCP/QUIC path that is only used for DTLS.

Reviewing my changes to `rustls/src/msgs/deframer` and `rustls/src/msgs/fragmenter.rs`, it does appear that I principally add entirely new functions for the DTLS paths. These could be refactored into distinct modules/items, something like `DtlsDeframer` and `DtlsFragmenter`. I look forward to discussing that in the PRs I'll open for those pieces.

> Again, this sounds like mayb eresuing the existing `Deframer` doesn't make sense (and maybe not even `ReceivePath`?).

I think there's a tricky balance to be struck here: I wouldn't want to over burden the existing `SendPath, ReceivePath` implementation with complicated conditional logic that is ignored by 99% of crate users who just do connection-oriented TLS, but I'm concerned that if we go too hard on the path of segregating the protocols, then we'll end up duplicating a lot of code, or spending time rediscovering bugs and corner cases in the DTLS path that have long since been fixed on the TLS path. To be clear, I'm not rejecting your point here and broadly I intend to defer to the judgement and preference of rustls maintainers. But I want to highlight a tension that I think we'll have to navigate here.

Relatedly, my approach thus far has been inspired by the QUIC implementation, where we see a number of runtime checks for QUIC affecting state machine decisions. If we go for an architecture that cleanly segregates DTLS from TLS, is there an appetite for doing something similar to QUIC?

> Timeouts and retransmission should be wrapped around the inner TLS state.

Does this mean you see it happening between the state machines (`client::{hs, tls12, tls13}`, `server::{hs, tls12, tls13}`) and `conn::{SendPath, ReceivePath}` (or whatever DTLS-specific equivalent to the latter we put together)?

> What is the base of your branch?

I started working from [`bcae7d8`](https://github.com/rustls/rustls/commit/bcae7d806ad737a532501db73cdefb8eb7ff9a6c), for no particular reason except that it was the tip of `main` when I started working on this. Given that lots of 0.24 work is in flight, I was always prepared to have to do some messy rebases or rearchitecting of this work. Even if I have to throw away a lot of what I've done, it was still worthwhile to deepen my own knowledge of rustls internals and the DTLS protocols.

> I think answering these questions [on validation and test plans] should be prioritized.

I completely agree. I'm going to invest time in studying other TLS implementations (like BoringSSL and NSS) to see how they approach this and whether there are test suites that we can either use directly (a la Bogo) or at least use to build a robust suite of test cases. I eagerly welcome more ideas for test strategy. 

> I think you answered this above? Probably 0.24 (but not necessarily 0.24.0).

Yes, that sounds right. This should manifest as public API additions, only, so it can be landed without breaking semver.

**Comment by divergentdave** [comment]
The `Read` and `Write` traits do not appropriately model a datagram socket, so I think we should eliminate those alternatives. Datagram-oriented protocols need to anchor their wire format on the start of each datagram, and readers and writers could easily lose track of the start of each datagram through buffering and splitting up read/write calls.

The DTLS implementation only needs to handle timeouts and retransmissions for handshake messages, not application data messages. Thus, I think these functions should probably be handled by the DTLS version of the handshake state machine, and not on the socket abstraction layer.

One further API design question: should the sans-I/O design principle be extended to the system clock as well? Unlike TLS, DTLS will need to set timers for handshake message retransmission. Externalizing the clock will certainly be helpful for testing purposes. When it comes to the public API, using the "sans-I/O for timers" approach might help because it could integrate with any of `std::time`, an async runtime, or `no_std` code, depending on the needs of the consumer.

**Comment by djc** [comment]
> > Note that I think `sendable_tls` is going to go away in favor of external buffers in 0.24. For DTLS, I'd aim for using external buffers from the start.
> 
> Ack. I will study the recent work on external buffers and adapt the implementation accordingly.

We haven't done the output buffers yet, but here's a PR for the input buffer:

- https://github.com/rustls/rustls/pull/2940

I expect changes to the send buffers will be similarish.

> > Again, this sounds like maybe reusing the existing `Deframer` doesn't make sense (and maybe not even `ReceivePath`?).
> 
> I think there's a tricky balance to be struck here: I wouldn't want to over burden the existing `SendPath, ReceivePath` implementation with complicated conditional logic that is ignored by 99% of crate users who just do connection-oriented TLS, but I'm concerned that if we go too hard on the path of segregating the protocols, then we'll end up duplicating a lot of code, or spending time rediscovering bugs and corner cases in the DTLS path that have long since been fixed on the TLS path. To be clear, I'm not rejecting your point here and broadly I intend to defer to the judgement and preference of rustls maintainers. But I want to highlight a tension that I think we'll have to navigate here.

Yup, I think we can tackle this over time, and potentially figure out new abstractions that enable sharing logic while keeping code paths similar.

> Relatedly, my approach thus far has been inspired by the QUIC implementation, where we see a number of runtime checks for QUIC affecting state machine decisions. If we go for an architecture that cleanly segregates DTLS from TLS, is there an appetite for doing something similar to QUIC?

IIRC the QUIC deviations inside the state machine are relatively limited, and I think we could do similar things for DTLS if necessary. But my impression is that most of the DTLS changes will be outside of the core state machine?

> > Timeouts and retransmission should be wrapped around the inner TLS state.
> 
> Does this mean you see it happening between the state machines (`client::{hs, tls12, tls13}`, `server::{hs, tls12, tls13}`) and `conn::{SendPath, ReceivePath}` (or whatever DTLS-specific equivalent to the latter we put together)?

Not exactly sure yet; see also below.

> > I think answering these questions [on validation and test plans] should be prioritized.
> 
> I completely agree. I'm going to invest time in studying other TLS implementations (like BoringSSL and NSS) to see how they approach this and whether there are test suites that we can either use directly (a la Bogo) or at least use to build a robust suite of test cases. I eagerly welcome more ideas for test strategy.

I think the first question is how much DTLS coverage there is in bogo.

> The `Read` and `Write` traits do not appropriately model a datagram socket, so I think we should eliminate those alternatives. Datagram-oriented protocols need to anchor their wire format on the start of each datagram, and readers and writers could easily lose track of the start of each datagram through buffering and splitting up read/write calls.

Agreed.

> The DTLS implementation only needs to handle timeouts and retransmissions for handshake messages, not application data messages. Thus, I think these functions should probably be handled by the DTLS version of the handshake state machine, and not on the socket abstraction layer.

On retransmission, do we rebuild the message or resend a cached message? If the latter, I don't think we want to hold those buffers inside the DTLS state.
 
> One further API design question: should the sans-I/O design principle be extended to the system clock as well? Unlike TLS, DTLS will need to set timers for handshake message retransmission. Externalizing the clock will certainly be helpful for testing purposes. When it comes to the public API, using the "sans-I/O for timers" approach might help because it could integrate with any of `std::time`, an async runtime, or `no_std` code, depending on the needs of the consumer.

quinn-proto takes `std::time::Instant` and never calls `Instant::now()` inside the library in order to be fully deterministic. That works well, but does tie it to `std` which we might want to avoid. rustls already has the `TimeProvider` trait, but that works with `UnixTime` which has second precision. That's probably not enough for this?

**Comment by ctz** [comment]
Thought I would write down how I would tackle this, in case that is useful.

1. First, I would be targetting https://www.ietf.org/archive/id/draft-rescorla-tls-rfc9147bis-00.html rather than RFC9147 itself for DTLS1.3. For DTLS1.2 I would review errata, mailing list chatter, other implementations, etc, to draw up a subset implementation target.

    I would suggest starting with DTLS1.3 rather than attacking both in parallel.

2. Next, I would be evaluating testing options. I suspect bogo has a good breadth of DTLS1.3 testing, but nothing for DTLS1.2. So another option would be needed here. If there isn't an independent option, one would be need to be written (in the style of bogo -- that is -- taking a working implementation and making it misbehave in a variety of interesting ways). In setting up testing, I would pay little attention to public API design, but I think functionally the IO-level needs to retain datagram boundaries rather than smush everything into a stream.

3. Next, I would "draw the rest of the owl" and smash out the whole implementation until all the tests pass. During this, I would be content to destroy the rest of the library in service of velocity; TCP-TLS would become a smoking wreck. The purpose of this is to learn a _complete_ set of changes required, rather than necessarily choose any particular path for each choice. But, importantly, we end up with effectively of a fork of rustls which can do DTLS1.x, with integration tests. This is a reachable goal in itself.

4. Now, for each unpalatable hack made in (3), I would retreat back to doing something more considered, and integrate this both upstream and into the "fork". For each iteration of this, the upstream (this repo) accretes new but as-yet unused capabilities (with no effect on TLS functionality, and hopefully code that is discarded at build-time) and the fork progressively gets closer to the upstream. At the end of this process, they are at parity, and nothing ever hits upstream that needs to change later (not that we can't do that, but exploring the design space via PRs landing on main is pretty expensive and slow).

5. With a rustls that does TCP-TLS1.2/1.3 and DTLS1.3 in the bag, I would follow the whole process again for  DTLS1.2. The set of changes should be smaller.

**Comment by ctz** [comment]
> * Should we have all this gated on a new cargo feature?

I would like to avoid any new use of cargo features that doesn't directly relate to control of dependencies.

**Comment by tgeoghegan** [comment]
Sorry for the delayed reply, I was on vacation last week.

> IIRC the QUIC deviations inside the state machine are relatively limited, and I think we could do similar things for DTLS if necessary. But my impression is that most of the DTLS changes will be outside of the core state machine?

Yes, that's true, but much like with QUIC, you'll see a little bit of DTLS specific logic, such as not changing ChangeCipherSpec messages for DTLS 1.3, or sending ACKs to handshake messages.

> On retransmission, do we rebuild the message or resend a cached message? If the latter, I don't think we want to hold those buffers inside the DTLS state.

It's the latter. Sending ACKs will be easy, I think: there's already plenty of places where the handshake state machines or the receive path can enqueue messages for sending. Receiving ACKs will be trickier.

What I envision is that the DTLS state layer will continually enqueue (usually encrypted) DTLS records for eventual transmission by the transmission layer, which will have to hold onto those records until they're ACKed by the other end. Once an ACK is received, it has to be deprotected and parsed by the DTLS state layer (or at least the receive path; point is the transmission layer can't do it) and then the RecordNumbers in the ACK have to be handed back to the transmission layer so that it can throw away the corresponding buffers. Since the buffers are encrypted, that means the transmission layer will need to keep track of which buffers contain which record numbers, but it should be fine to have the state layer convey that information along with the records.

> [Dirkjan] I think the first question is how much DTLS coverage there is in bogo.
> [Joe] I suspect bogo has a good breadth of DTLS1.3 testing, but nothing for DTLS1.2.

It looks like there's quite a lot of DTLS 1.2 coverage in BoGo's [dtls_tests.go](https://github.com/google/boringssl/blob/main/ssl/test/runner/dtls_tests.go), with tests being run against both protocol versions, using version specific logic where necessary (e.g., TLS 1.2 doesn't do ACKs). It's hard to say whether that provides sufficient coverage, though, but it sounds like it's good enough for BoringSSL. That said I'm looking into what other TLS implementations like NSS do to validate their DTLS implementations and will report back about the viability of importing their test suites. Then again, maybe they just run BoGo...

On how to move forward:

What I'm hearing from @djc is that he'd like to see PRs against main sooner rather than later so we don't get mired in a fork. But then @ctz is saying essentially the opposite, arguing for getting something robust working in a fork before upstreaming it piece-by-piece to rustls proper. I can see the merits of both approaches, and I'm happy to do either. I think we need the rustls maintainers to agree on how to proceed here.

**Comment by djc** [comment]
> What I'm hearing from [@djc](https://github.com/djc) is that he'd like to see PRs against main sooner rather than later so we don't get mired in a fork. But then [@ctz](https://github.com/ctz) is saying essentially the opposite, arguing for getting something robust working in a fork before upstreaming it piece-by-piece to rustls proper. I can see the merits of both approaches, and I'm happy to do either. I think we need the rustls maintainers to agree on how to proceed here.

Maybe get something working in a public draft PR so we can give early feedback?

**Comment by tgeoghegan** [comment]
OK, sounds good. Early next week I will:

- Post a draft PR containing my WIP so architectural feedback can start
- Emit a PR against current `main` that has just the DTLS fragmenter, which I believe could be taken with minimal disruption to TCP TLS.

**Comment by tgeoghegan** [comment]
A little later than I wanted but I have posted a draft PR: #3110 

Besides continuing discussion there, my next step would be to make a series of smaller PRs that:

- introduce an `EncodingContext` field to `OpaqueOutbound` and push decisions like which protocol version to claim down to message encoding (#3112)
- add message definitions with encoders and decoders for various DTLS messages
- add DTLS support to the message fragmenter
- add DTLS support to the message deframer+coalescer
